File: ahc/39/39.rs
use proconio::input;
use rand::Rng;
use std::time::Instant;

fn gen_rec() -> Vec<(i32, i32)> {
    let mut rng = rand::thread_rng();
    loop {
        let x = rng.gen_range(0..=100_000);
        let y = rng.gen_range(0..=100_000);
        let width = rng.gen_range(500..=100_000);
        let height = rng.gen_range(500..=100_000);

        if x >= width && y >= height {
            return vec![
                (x, y),
                (x, y - height),
                (x - width, y - height),
                (x - width, y),
            ];
        }
    }
}
fn is_point_inside_polygon(maxx: i32, minx: i32, maxy: i32, miny: i32, x: i32, y: i32) -> bool {
    minx <= x && x <= maxx && miny <= y && y <= maxy
}

fn main() {
    input! {
        n: u16,
        sabas:[(i32, i32); n],
        iwashis:[(i32, i32); n],
    }

    let time_limit = 1.98;
    let start_time = Instant::now();

    let mut max_cnt = -1;
    let mut max_rec: Vec<(i32, i32)> = Vec::new();

    while start_time.elapsed().as_secs_f64() < time_limit {
        let mut cnt = 0;
        let rec = gen_rec();

        let mut maxx = -1;
        let mut minx = i32::MAX;
        let mut maxy = -1;
        let mut miny = i32::MAX;

        for &(x, y) in &rec {
            maxx = std::cmp::max(maxx, x);
            minx = std::cmp::min(minx, x);
            maxy = std::cmp::max(maxy, y);
            miny = std::cmp::min(miny, y);
        }

        for &(x, y) in &sabas {
            if is_point_inside_polygon(maxx, minx, maxy, miny, x, y) {
                cnt += 1;
            }
        }

        for &(x, y) in &iwashis {
            if is_point_inside_polygon(maxx, minx, maxy, miny, x, y) {
                cnt -= 1;
            }
        }

        if cnt > max_cnt {
            max_cnt = cnt;
            max_rec = rec.clone();
        }
    }

    println!("{}", max_rec.len());
    for &(x, y) in &max_rec {
        println!("{} {}", x, y);
    }
}

File: ahc/39/39.py
import random
import time

N = int(input())
sabas = [tuple(map(int, input().split())) for _ in range(N)]
iwashis = [tuple(map(int, input().split())) for _ in range(N)]
TIME_LIMIT = 1.98


def gen_rec():
    while True:
        x, y = random.randint(0, 10**5), random.randint(0, 10**5)
        width, height = random.randint(500, 10**5), random.randint(500, 10**5)
        if x - width >= 0 and y - height >= 0:
            return [(x, y), (x, y - height), (x - width, y - height), (x - width, y)]


def is_point_inside_polygon(maxx, minx, maxy, miny, x, y):
    if minx <= x <= maxx and miny <= y <= maxy:
        return 1
    else:
        return 0


def main():
    start_time = time.time()
    max_cnt = -1
    max_rec = None
    while time.time() - start_time < TIME_LIMIT:
        cnt = 0
        rec = gen_rec()
        maxx, minx = -1, float("inf")
        maxy, miny = -1, float("inf")
        for x, y in rec:
            maxx = max(maxx, x)
            minx = max(minx, x)
            maxy = max(maxy, y)
            miny = max(miny, y)
        for saba in sabas:
            if is_point_inside_polygon(maxx, minx, maxy, miny, saba[0], saba[1]):
                cnt += 1
        for iwashi in iwashis:
            if is_point_inside_polygon(maxx, minx, maxy, miny, iwashi[0], iwashi[1]):
                cnt -= 1
        if cnt > max_cnt:
            max_cnt = cnt
            max_rec = rec

    print(len(max_rec))
    for r in max_rec:
        print(*r)


if __name__ == "__main__":
    main()
